**Drive'ı bağlıyoruz.**

In [1]:
# prompt: add drive mount code

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


**Ekran kartını, ekran kartı hafızasını, Cuda versiyonunu, driver versiyonunu kontrol ediyoruz.**

In [ ]:
!nvidia-smi

Sun Dec 22 20:21:23 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A100-SXM4-40GB          Off | 00000000:00:04.0 Off |                    0 |
| N/A   29C    P0              46W / 400W |      2MiB / 40960MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+--

**Veri setini Kaggle'dan yüklüyoruz.**

In [ ]:
# IMPORTANT: RUN THIS CELL IN ORDER TO IMPORT YOUR KAGGLE DATA SOURCES,
# THEN FEEL FREE TO DELETE THIS CELL.
# NOTE: THIS NOTEBOOK ENVIRONMENT DIFFERS FROM KAGGLE'S PYTHON
# ENVIRONMENT SO THERE MAY BE MISSING LIBRARIES USED BY YOUR
# NOTEBOOK.
import kagglehub
rrebirrth_animals_with_attributes_2_path = kagglehub.dataset_download('rrebirrth/animals-with-attributes-2')

print('Data source import complete.')

100%|██████████| 13.0G/13.0G [01:02<00:00, 224MB/s]

Extracting files...


Data source import complete.


# Drive da çalışmak için kullanılacak

**!!**: *Kendi drive dizininizi yazın.*

In [2]:
# Dizin kontrolü
%pwd

'/content'

In [ ]:
# Drive dizininin içeriği
%ls /content/drive/MyDrive/bootcamp/animal_classification/

config/  README.md  requirements.txt  src/


In [ ]:
#dizini değiştiriyoruz, drive da çalışalım: /content/drive/MyDrive/bootcamp/animal_classification/
%cd /content/drive/MyDrive/bootcamp/animal_classification

/content/drive/MyDrive/bootcamp/animal_classification


**Veri setinden sadece kullanılacak sınıfların ilk 650 görüntüsünü dizine kaydediyoruz.**

---

***Sınıflar:***

- *collie*
- *dolphin*
- *elephant*
- *fox*
- *moose*
- *rabbit*
- *sheep*
- *squirrel*
- *giant+panda*
- *polar+bear*

In [ ]:
import os
import shutil
from PIL import Image

# Initialize the dictionary to hold paths for each animal
image_paths = {}

# Base directory for your dataset
base_path = "/root/.cache/kagglehub/datasets/rrebirrth/animals-with-attributes-2/versions/1/Animals_with_Attributes2/JPEGImages/"

# List of animals to search for
animals = ["collie", "dolphin", "elephant", "fox", "moose", "rabbit", "sheep", "squirrel", "giant+panda", "polar+bear"]

# Traverse the directory structure to collect image paths
for dirname, _, filenames in os.walk(base_path):
    for animal in animals:
        if animal in dirname:
            if animal not in image_paths:
                image_paths[animal] = []
            for filename in filenames:
                image_paths[animal].append(os.path.join(dirname, filename))

datasets_dir = "dataset/raw"
# Create the 'dataset/raw' directory if it doesn't exist
os.makedirs(datasets_dir, exist_ok=True)

# Copy the first 650 images for each animal
for animal, paths in image_paths.items():
    # Create a directory for the current animal within 'dataset/raw'
    animal_dir = os.path.join(datasets_dir, animal)
    os.makedirs(animal_dir, exist_ok=True)

    # Copy the first 650 images
    for i, path in enumerate(paths[:650]):
        destination_path = os.path.join(animal_dir, os.path.basename(path))
        shutil.copy(path, destination_path)

print("Images copied successfully!")

# Get a list of all class names (subdirectories) within the base directory
class_names = [f for f in os.listdir(datasets_dir) if os.path.isdir(os.path.join(datasets_dir, f))]

# Loop through each class name and count the images
for class_name in class_names:
    # Construct the path to the current class's directory
    class_dir = os.path.join(datasets_dir, class_name)

    # Initialize counters for total and empty images
    total_images = 0
    empty_images = 0

    # Loop through files in the class directory
    for filename in os.listdir(class_dir):
        # Construct the full file path
        file_path = os.path.join(class_dir, filename)

        # Check if it's a file (not a subdirectory)
        if os.path.isfile(file_path):
            total_images += 1

            # Attempt to open the image using Pillow
            try:
                with Image.open(file_path) as img:
                    # Check if the image is empty (size is 0)
                    if img.size == (0, 0):
                        empty_images += 1
            except IOError:
                # Handle cases where the file is not a valid image
                print(f"Warning: Could not open or read image file: {file_path}")

    # Print the class name, total images, and empty image count
    print(f"{class_name}: {total_images} images, {empty_images} empty")

Images copied successfully!
giant+panda: 650 images, 0 empty
polar+bear: 650 images, 0 empty
rabbit: 650 images, 0 empty
elephant: 650 images, 0 empty
collie: 650 images, 0 empty
squirrel: 650 images, 0 empty
sheep: 650 images, 0 empty
fox: 650 images, 0 empty
moose: 650 images, 0 empty
dolphin: 650 images, 0 empty


# Colab `/content` dizininde çalışmak için kullanılacak

In [3]:
%pwd

'/content'

In [4]:
!git clone https://github.com/dijvar/CNN-Animal-Classification.git

Cloning into 'CNN-Animal-Classification'...
remote: Enumerating objects: 34, done.
remote: Counting objects: 100% (34/34), done.
remote: Compressing objects: 100% (30/30), done.
remote: Total 34 (delta 12), reused 18 (delta 3), pack-reused 0 (from 0)
Receiving objects: 100% (34/34), 71.37 KiB | 4.20 MiB/s, done.
Resolving deltas: 100% (12/12), done.


In [5]:
%cd CNN-Animal-Classification

/content/CNN-Animal-Classification


**Veri setinden sadece kullanılacak sınıfların ilk 650 görüntüsünü dizine kaydediyoruz.**

---

***Sınıflar:***

- *collie*
- *dolphin*
- *elephant*
- *fox*
- *moose*
- *rabbit*
- *sheep*
- *squirrel*
- *giant+panda*
- *polar+bear*

In [ ]:
import os
import shutil
from PIL import Image

# Initialize the dictionary to hold paths for each animal
image_paths = {}

# Base directory for your dataset
base_path = "/root/.cache/kagglehub/datasets/rrebirrth/animals-with-attributes-2/versions/1/Animals_with_Attributes2/JPEGImages/"

# List of animals to search for
animals = ["collie", "dolphin", "elephant", "fox", "moose", "rabbit", "sheep", "squirrel", "giant+panda", "polar+bear"]

# Traverse the directory structure to collect image paths
for dirname, _, filenames in os.walk(base_path):
    for animal in animals:
        if animal in dirname:
            if animal not in image_paths:
                image_paths[animal] = []
            for filename in filenames:
                image_paths[animal].append(os.path.join(dirname, filename))

datasets_dir = "dataset/raw"
# Create the 'dataset/raw' directory if it doesn't exist
os.makedirs(datasets_dir, exist_ok=True)

# Copy the first 650 images for each animal
for animal, paths in image_paths.items():
    # Create a directory for the current animal within 'dataset/raw'
    animal_dir = os.path.join(datasets_dir, animal)
    os.makedirs(animal_dir, exist_ok=True)

    # Copy the first 650 images
    for i, path in enumerate(paths[:650]):
        destination_path = os.path.join(animal_dir, os.path.basename(path))
        shutil.copy(path, destination_path)

print("Images copied successfully!")

# Get a list of all class names (subdirectories) within the base directory
class_names = [f for f in os.listdir(datasets_dir) if os.path.isdir(os.path.join(datasets_dir, f))]

# Loop through each class name and count the images
for class_name in class_names:
    # Construct the path to the current class's directory
    class_dir = os.path.join(datasets_dir, class_name)

    # Initialize counters for total and empty images
    total_images = 0
    empty_images = 0

    # Loop through files in the class directory
    for filename in os.listdir(class_dir):
        # Construct the full file path
        file_path = os.path.join(class_dir, filename)

        # Check if it's a file (not a subdirectory)
        if os.path.isfile(file_path):
            total_images += 1

            # Attempt to open the image using Pillow
            try:
                with Image.open(file_path) as img:
                    # Check if the image is empty (size is 0)
                    if img.size == (0, 0):
                        empty_images += 1
            except IOError:
                # Handle cases where the file is not a valid image
                print(f"Warning: Could not open or read image file: {file_path}")

    # Print the class name, total images, and empty image count
    print(f"{class_name}: {total_images} images, {empty_images} empty")

Images copied successfully!
giant+panda: 650 images, 0 empty
polar+bear: 650 images, 0 empty
rabbit: 650 images, 0 empty
elephant: 650 images, 0 empty
collie: 650 images, 0 empty
squirrel: 650 images, 0 empty
sheep: 650 images, 0 empty
fox: 650 images, 0 empty
moose: 650 images, 0 empty
dolphin: 650 images, 0 empty


# Genel akış


In [ ]:
# Tensorflow sürümünü kontrol ediyoruz.
!pip show tensorflow
!pip show tensorflow-gpu

Name: tensorflow
Version: 2.17.1
Summary: TensorFlow is an open source machine learning framework for everyone.
Home-page: https://www.tensorflow.org/
Author: Google Inc.
Author-email: packages@tensorflow.org
License: Apache 2.0
Location: /usr/local/lib/python3.10/dist-packages
Requires: absl-py, astunparse, flatbuffers, gast, google-pasta, grpcio, h5py, keras, libclang, ml-dtypes, numpy, opt-einsum, packaging, protobuf, requests, setuptools, six, tensorboard, tensorflow-io-gcs-filesystem, termcolor, typing-extensions, wrapt
Required-by: dopamine_rl, tf_keras


In [ ]:
# Tensorflow'un GPU kullanımını kontrol ediyoruz.
import tensorflow as tf
device_name = tf.test.gpu_device_name()
if device_name != '/device:GPU:0':
  raise SystemError('GPU device not found')
print('Found GPU at: {}'.format(device_name))

Found GPU at: /device:GPU:0


## Model Eğitimi ve Değerlendirmesi

Bu komut, CNN modelinin eğitimini ve değerlendirmesini başlatır. Eğitim süreci aşağıdaki adımları içerir:

1. **Veri Seti Hazırlama**
   - 10 sınıf için toplam 6500 görüntü (her sınıf için 650)
   - Train (%70), Validation (%20), Test (%10) olarak bölünme
   - Görüntü boyutları 224x224 ve 128x128 olarak yeniden boyutlandırma
   - Piksel değerleri [0,1] aralığına normalizasyon

2. **Grid Search Parametreleri**
   - Input Size: [224, 128]
   - Aktivasyon Fonksiyonları: [ReLU, PReLU, ELU]
   - Learning Rate: [0.01, 0.001, 0.0001]
   - Data Augmentation: [True, False]
   - Toplam 36 farklı model konfigürasyonu test edilecek

3. **Model Mimarisi**
   - 3 Konvolüsyon Bloğu (32-64-128 filtre)
   - Batch Normalization
   - MaxPooling
   - Dropout (0.2)
   - Dense Layer (256 nöron)
   - L2 Regularization (weight_decay=0.0001)

4. **Eğitim Detayları**
   - Batch Size: 64
   - Epochs: 100
   - Early Stopping (patience=10)
   - Model Checkpoint (en iyi validation accuracy)
   - Wandb ile deney takibi

5. **Test Senaryoları**
   - Orijinal test görüntüleri
   - Farklı ışık koşullarında manipüle edilmiş görüntüler
   - Renk sabitliği algoritması uygulanmış görüntüler

6. **Çıktılar**
   - Her deney için model performans metrikleri
   - Hücre çıktısının sonunda en başarılı deney yazdırılmıştır
   - Test sonuçları (accuracy ve loss değerleri)
   - En başarılı model konfigürasyonu
   - [Wandb](https://wandb.ai/orhandijvar/bootcamp-cnn-animal-classification/table?nw=nwuserorhandijvar) üzerinden detaylı analiz grafikleri

**!!:** Eğitim süreci yaklaşık 8-10 saat sürebilir (GPU kullanımına bağlı olarak).

*NVIDIA A100-40GB GPU ile yaklaşık 4 saat sürmüştür.*  

In [ ]:
!python src/train.py

Streaming output truncated to the last 5000 lines.
wandb: Synced 5 W&B file(s), 0 media file(s), 0 artifact file(s) and 0 other file(s)
wandb: Find logs at: results/wandb/wandb/run-20241222_222725-gb9b1k0u/logs

Deney 12/36 tamamlandı.

==================== Deney 13/36 ====================
Deney Adı: exp_size_224_act_elu_lr_0.01_aug_True
Input Size: 224
Activation: elu
Learning Rate: 0.01
Data Augmentation: Aktif
wandb: Tracking run with wandb version 0.19.1
wandb: Run data is saved locally in results/wandb/wandb/run-20241222_222916-xxqy6iw6
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run exp_size_224_act_elu_lr_0.01_aug_True
wandb: ⭐️ View project at https://wandb.ai/orhandijvar/bootcamp-cnn-animal-classification
wandb: 🚀 View run at https://wandb.ai/orhandijvar/bootcamp-cnn-animal-classification/runs/xxqy6iw6

Model Mimarisi:
--------------------------------------------------
Model: "functional"
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳